database2/ffboltdb/db.go

```diff
@@ -0,0 +1,1510 @@
+// provides a root bucket against which all read and writes occur.
+type transaction struct {
+	managed        bool     // Is the transaction managed?
+	closed         bool     // Is the transaction closed?
+	writable       bool     // Is the transaction writable?
+	db             *db      // DB instance the tx was created from.
+	boltTx         *bolt.Tx // Underlying bolt tx for metadata storage.
+	metaBucket     *bucket  // The metadata bucket in underlying bolt DB.
+	blockIdxBucket *bucket  // The block index bucket.
+
+	// Blocks that need to be stored on commit.  The pendingBlocks map is
+	// kept to allow quick looks up pending data by block hash.
+	pendingBlocks    map[wire.ShaHash]int
+	pendingBlockData []pendingBlock
+}
+
+// Enforce transaction implements the database.Tx interface.
+var _ database.Tx = (*transaction)(nil)
+
+// checkClosed returns an error if the the database or transaction is closed.
+func (tx *transaction) checkClosed() error {
+	// The transaction is no longer valid if it has been closed.
+	if tx.closed {
+		return makeDbErr(database.ErrTxClosed, errTxClosedStr, nil)
+	}
+
+	return nil
+}
+
+// Metadata returns the top-most bucket for all metadata storage.
+//
+// This function is part of the database.Tx interface implementation.
+func (tx *transaction) Metadata() database.Bucket {
+	return tx.metaBucket
+}
+
+// hasBlock returns whether or not a block with the given hash exists.
+func (tx *transaction) hasBlock(hash *wire.ShaHash) bool {
+	// Return true if the block is pending to be written on commit since
+	// it exists from the viewpoint of this transaction.
+	if tx.pendingBlocks != nil {
+		if _, exists := tx.pendingBlocks[*hash]; exists {
+			return true
+		}
+	}
+
+	// Bolt is zero-copy so this doesn't incur additional overhead of
+	// loading the entry.
+	return tx.blockIdxBucket.Get(hash[:]) != nil
+}
+
+// StoreBlock stores the provided block into the database.  There are no checks
+// to ensure the block connects to a previous block, contains double spends, or
+// any additional functionality such as transaction indexing.  It simply stores
+// the block in the database.
+//
+// Returns the following errors as required by the interface contract:
+//   - ErrBlockExists when the block hash already exists
+//   - ErrTxNotWritable if attempted against a read-only transaction
+//   - ErrTxClosed if the transaction has already been closed
+//
+// This function is part of the database.Tx interface implementation.
+func (tx *transaction) StoreBlock(block *btcutil.Block) error {
+	// Ensure transaction state is valid.
+	if err := tx.checkClosed(); err != nil {
+		return err
+	}
+
+	// Ensure the transaction is writable.
+	if !tx.writable {
+		str := "store block requires a writable database transaction"
+		return makeDbErr(database.ErrTxNotWritable, str, nil)
+	}
+
+	// Reject the block if it already exists.
+	blockHash := block.Sha()
+	if tx.pendingBlocks != nil {
```

Comment:
Right.  It was there before because I was keeping the index to know when to replace the block later, but I forget to remove this check when I removed that flag.